domain/blockdevice/state/state.go

```diff
@@ -0,0 +1,401 @@
+	if err != nil {
+		return nil, errors.Annotatef(err, "loading block devices for machine %q", machineId)
+	}
+	return dbRows.toBlockDevices(dbDeviceLinks, dbFilesystemTypes)
+}
+
+// GetMachineInfo is used look up the machine UUID and life for a machine.
+func (st *State) GetMachineInfo(ctx context.Context, machineId string) (string, domain.Life, error) {
+	db, err := st.DB()
+	if err != nil {
+		return "", 0, errors.Trace(err)
+	}
+
+	var (
+		machineUUID string
+		life        domain.Life
+	)
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		var err error
+		machineUUID, life, err = getMachineInfo(ctx, tx, machineId)
+		return errors.Trace(err)
+	})
+	return machineUUID, life, errors.Trace(err)
+}
+
+func getMachineInfo(ctx context.Context, tx *sqlair.TX, machineId string) (string, domain.Life, error) {
+	q := `
+SELECT machine.life_id AS &M.life_id, machine.uuid AS &M.machine_uuid
+FROM   machine
+WHERE  machine.machine_id = $M.machine_id
+`
+	stmt, err := sqlair.Prepare(q, sqlair.M{})
+	if err != nil {
+		return "", 0, errors.Trace(err)
+	}
+
+	result := sqlair.M{}
+	err = tx.Query(ctx, stmt, sqlair.M{"machine_id": machineId}).Get(result)
+	if err != nil && !errors.Is(err, sqlair.ErrNoRows) {
+		return "", 0, errors.Trace(err)
+	}
+	if len(result) == 0 {
+		return "", 0, fmt.Errorf("machine %q %w", machineId, errors.NotFound)
+	}
+	life, ok := result["life_id"].(int64)
+	if !ok {
+		return "", 0, errors.Errorf("missing life value for machine %q", machineId)
+	}
+	machineUUID := result["machine_uuid"].(string)
+	return machineUUID, domain.Life(life), nil
+}
+
+// SetMachineBlockDevices sets the block devices visible on the machine.
+// Previously recorded block devices not in the list will be removed.
+func (st *State) SetMachineBlockDevices(ctx context.Context, machineId string, devices ...blockdevice.BlockDevice) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		machineUUID, life, err := getMachineInfo(ctx, tx, machineId)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if life == domain.Dead {
+			return errors.Errorf("cannot update block devices on dead machine %q", machineId)
+		}
+		existing, err := loadBlockDevices(ctx, tx, machineId)
+		if err != nil {
+			return errors.Annotatef(err, "loading block devices for machine %q", machineId)
+		}
+		if !blockDevicesChanged(existing, devices) {
+			return nil
+		}
+
+		if err := updateBlockDevices(ctx, tx, machineUUID, devices...); err != nil {
+			return errors.Annotatef(err, "updating block devices on machine %q (%s)", machineId, machineUUID)
+		}
+		return nil
+	})
+
+	return errors.Trace(err)
+}
+
+func updateBlockDevices(ctx context.Context, tx *sqlair.TX, machineUUID string, devices ...blockdevice.BlockDevice) error {
+	fsTypeQuery := `SELECT * AS &FilesystemType.* FROM filesystem_type`
```

Comment:
Why can't we just select this out as string, int into a map with sqlair. SQL is there to get it into the form we want straight away without the need for swapping over to another type struct in code.

Reply:
We could select into a map but I prefer an explicit struct. Personal preference. Avoids type conversions when reading the values from the map etc.